In [ ]:
import torch
import numpy as np
import matplotlib.pyplot as plt
import os
import sys
import shutil
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as transforms
import torch.nn.functional as F
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Define source and destination directories
source_dir = "dir/Diffusion/results/train_1_2/"
output_dir = "dir/Diffusion/results/train_1_2/100png"

# Ensure output directory exists
os.makedirs(output_dir, exist_ok=True)

# Iterate through folders named sample_0 to sample_99
for i in range(100):
    folder_name = f"sample_{i}"
    png_folder_path = os.path.join(source_dir, folder_name, "png")
    image_path = os.path.join(png_folder_path, f"{i}_0.png")

    if os.path.exists(image_path):
        output_filename = f"sample_{i}_0_0.png"
        shutil.copy(image_path, os.path.join(output_dir, output_filename))
        print(f"Copied {image_path} to {output_dir} as {output_filename}")
    else:
        print(f"{image_path} not found!")

print("Extraction complete!")



Copied /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/sample_0/png/0_0.png to /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/100png as sample_0_0_0.png
Copied /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/sample_1/png/1_0.png to /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/100png as sample_1_0_0.png
Copied /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/sample_2/png/2_0.png to /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/100png as sample_2_0_0.png
Copied /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/sample_3/png/3_0.png to /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/100png as sample_3_0_0.png
Copied /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/sample_4/png/4_0.png to /content/drive/MyDrive/Hari/Evaluation/Diffusion/results/train_1_2/100png as sample_4_0_0.png
Copied /content/drive/MyDrive/Hari/Evalu

In [ ]:
import os
import random
import torch
import numpy as np
import torchvision.transforms as transforms
from PIL import Image

# Paths
source_folder = "dir/Diffusion/data/Samples"
destination_folder = "dir/Diffusion/data/downscaled_100_pngs"
#downscale_size = (28)  = > Desired downscaled size

# Making the directory:
os.makedirs(destination_folder, exist_ok=True)

# Get all .npy files
npy_files = [f for f in os.listdir(source_folder) if f.endswith(".npy")]

# Check if there are enough files to sample
if len(npy_files) < 100:
    print("Not enough .npy files in the source folder!")
else:
    selected_files = random.sample(npy_files, 100)  # Randomly pick 100 .npy files

    for file in selected_files:
        # Load .npy file
        npy_path = os.path.join(source_folder, file)
        npy_data = np.load(npy_path)  # Shape assumed to be (1, H, W) for grayscale

        # Convert to PyTorch tensor
        tensor = torch.tensor(npy_data, dtype=torch.float32)  # Shape: (1, H, W)

        # Add batch dimension for interpolation: (B, C, H, W)
        tensor = tensor.unsqueeze(0)  # Shape: (1, 1, H, W)

        # Downscale using interpolate
        tensor_downscaled = torch.nn.functional.interpolate(
            tensor, size=(28), mode="bilinear", align_corners=False
        )

        # Remove batch dimension: (1, 28, 28) -> (28, 28)
        tensor_downscaled = tensor_downscaled.squeeze(0).squeeze(0)

        # Convert to PIL image
        transform_to_pil = transforms.ToPILImage()
        image = transform_to_pil(tensor_downscaled)

        # Save as PNG
        png_filename = os.path.splitext(file)[0] + ".png"  # Change extension
        image.save(os.path.join(destination_folder, png_filename))

    print("Successfully processed and saved 100 downscaled PNGs!")


Successfully processed and saved 100 downscaled PNGs!


In [ ]:
!pip install pytorch-fid

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 107.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 84.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 55.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 17.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 95.1 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling nvidia-nvjitlink-cu12-12.5.82:
      Successfully uninstalled nvidia-nvjitli

In [ ]:
#Official FID implementation  : carrying out on 100 pngs
from pytorch_fid import fid_score

path_real = "dir/Diffusion/data/downscaled_100_pngs"
path_gen = "dir/Diffusion/results/train_1_2/100png"

fid_value = fid_score.calculate_fid_given_paths([path_real, path_gen], batch_size=10, device="cpu", dims=2048)
print(f"FID Score: {fid_value}")

Downloading: "https://github.com/mseitzer/pytorch-fid/releases/download/fid_weights/pt_inception-2015-12-05-6726825d.pth" to /root/.cache/torch/hub/checkpoints/pt_inception-2015-12-05-6726825d.pth
100%|██████████| 91.2M/91.2M [00:00<00:00, 443MB/s]
100%|██████████| 10/10 [00:22<00:00,  2.28s/it]


FID Score: 27.91234606780995


**FID = 27.91**

**As one can notice we obtained a much lower FID score as compared to the previous analysis, this is because I ran FID on the natural dimension of the generated images and the dimension of the train images was bought to what the model had seen. I can expect an even lower FID if I use a bigger sample size. I used a nominal size of 100, since I already had 100 samples of generated images in natural 28 spatial dimension , all other generated image samples I have at hand are upscaled to 150. And I do not have the Colab GPU access as of now to generate more samples.**

The model weights I submit are stored at  "./results/model_checkpoint.pth"